packages/eslint-plugin/src/rules/consistent-return.ts

```diff
@@ -0,0 +1,95 @@
+import { type TSESTree } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import type {
+  InferMessageIdsTypeFromRule,
+  InferOptionsTypeFromRule,
+} from '../util';
+import { createRule, getParserServices, isTypeFlagSet } from '../util';
+import { getESLintCoreRule } from '../util/getESLintCoreRule';
+
+const baseRule = getESLintCoreRule('consistent-return');
+
+type Options = InferOptionsTypeFromRule<typeof baseRule>;
+type MessageIds = InferMessageIdsTypeFromRule<typeof baseRule>;
+
+type FunctionNode =
+  | TSESTree.FunctionDeclaration
+  | TSESTree.FunctionExpression
+  | TSESTree.ArrowFunctionExpression;
+
+export default createRule<Options, MessageIds>({
+  name: 'consistent-return',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description:
+        'Require `return` statements to either always or never specify values',
+      extendsBaseRule: true,
+      requiresTypeChecking: true,
+    },
+    hasSuggestions: baseRule.meta.hasSuggestions,
+    schema: baseRule.meta.schema,
+    messages: baseRule.meta.messages,
+  },
+  defaultOptions: [],
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+    const rules = baseRule.create(context);
+
+    let functionNode: FunctionNode | null;
```

Comment:
I think we should store function scopes as stack (`FunctionDeclaration` - push on top, `FunctionDeclaration:exit` - pop from top), because 

This doesn't report any errors ([playground](https://deploy-preview-8289--typescript-eslint.netlify.app/play#ts=5.3.3&fileType=.tsx&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXwCNYAKASgC54A3HLYAbgCgV1s8kcdjFoBzSgpwggoqCtVrB4Ab0bx4WRPG59SMufPhwMyGPiIwyTeQF8N23aiYmgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6MgeyeUuX0Ra1oifLGhN0URNGgdokcGAC%2BIJUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)):

```ts
declare function bar(): void;
function foo(flag: boolean): void {
  if (flag) {
    return bar();
  }
  return;
}
```

But this does report ([playground](https://deploy-preview-8289--typescript-eslint.netlify.app/play#ts=5.3.3&fileType=.tsx&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXwCNYAKASgC54A3HLYAbgCgV1s8kcdjFoBzSgpwggoqCtVrB4Ab0bwkaTLnyJUZSgGcMMLKl4y58%2BHAzIY%2BAOQBGC03kBfQ1kTxufUgaPGQp84RKkdvCO8iZmqEz2QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6MgeyeUuX0Ra1oifLGhN0URNGgdokcGAC%2BIJUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)):

```ts
declare function bar(): void;
function foo(flag: boolean): void {
  function fn(): string {
    return '1';
  }
  if (flag) {
    return bar();
  }
  return;
}
```

I just added `function fn(): string` which does nothing, but gives `functionNode` a new value, so the return statements from the upper scope functions think they are inside the `fn` function.

Reply:
@auvred Oj, I completely missed nested cases, thanks for the good point.